sungc/deglint.py

```diff
@@ -30,7 +28,552 @@
+        }
+
+        xr_dvars = {}
+        for visname in vis_bands:
+            vis_im = self.xarr.variables[visname].values  # 3D
+            nodata = self.xarr.variables[visname].attrs["nodata"]
+            dtype = self.xarr.variables[visname].dtype
+
+            # create empty output array
+            deglinted_im = np.zeros(
+                [self.ntime, self.nrows, self.ncols], order="C", dtype=dtype
+            )
+
+            success_bands = []
+
+            # iterate through time
+            for t in range(self.ntime):
+
+                # check for nodata/empty arrays:
+                if rf.check_singleval(vis_im[t, :, :], nodata) or rf.check_singleval(
+                    corr_im[t, :, :], nodata
+                ):
+                    deglinted_im[t, :, :] = np.full((self.nrows, self.ncols), nodata)
+                    success_bands.append(False)
+                    continue  # go to next timestamp
+
+                # water_mask should identify all valid water pixels
+                water_mask = (vis_im[t, :, :] != nodata) & (corr_im[t, :, :] != nodata)
+
+                # create mask from the shapefile of this timestamp
+                roi_mask = rf.load_mask_from_shp(roi_shplist[t], metad)
+                roi_mask[~water_mask] = False
+
+                if plot:
+                    vis_bname = f"{visname}_time{t}"
+                    corr_bname = f"{corr_band}_time{t}"
+                    plot_tuple = (fig, ax, vis_bname, corr_bname, odir)
+                else:
+                    plot_tuple = None
+
+                deglinted_im[t, :, :], success = hedley_backend(
+                    vis_im=vis_im[t, :, :],
+                    corr_im=corr_im[t, :, :],
+                    water_mask=water_mask,
+                    roi_mask=roi_mask,
+                    nodata=nodata,
+                    scale_factor=self.scale_factor,
+                    clip=False,
+                    plot=plot,
+                    plot_tuple=plot_tuple,
+                )
+
+                success_bands.append(success)
+
+            # success_bands indicate which band/timestamps that the
+            # deglinting failed - typically as a result of an
+            # insufficient number of valid pixels in the ROI
+            xr_dvars[f"{visname}"] = xr.Variable(
+                dims=["time", "y", "x"],
+                data=deglinted_im,
+                attrs={"deglint_success": success_bands},
+            )
+
+        # create the output xarray.Dataset
+        new_attrs = self.xarr.attrs.copy()
+        new_attrs["deglint_algorithm"] = "Hedley et al. (2005)"
+
+        new_coords = {
+            "time": self.xarr.coords["time"].values,
+            "y": self.xarr.coords["y"].values,
+            "x": self.xarr.coords["x"].values,
+        }
+
+        deglinted_dxr = xr.Dataset(data_vars=xr_dvars, coords=new_coords, attrs=new_attrs)
+
+        # cleanup tmpdir
+        if isinstance(tmpdir, tempfile.TemporaryDirectory):
```

Comment:
You're assigning tmpdir to be either None or an instance of TemporaryDirectory.

It would be simpler to check for None as that is your default:

```suggestion
        if not (tmpdir is None):
```